odxtools/compumethods/scalelinearcompumethod.py

```diff
@@ -1,39 +1,95 @@
-        return lin_method.convert_physical_to_internal(physical_value)
-
-    def convert_internal_to_physical(self, internal_value: AtomicOdxType) -> AtomicOdxType:
-        lin_method = next(
-            scale for scale in self.linear_methods if scale.is_valid_internal_value(internal_value))
-        return lin_method.convert_internal_to_physical(internal_value)
+    def segments(self) -> List[LinearSegment]:
+        return self._segments
+
+    @staticmethod
+    def compu_method_from_et(et_element: ElementTree.Element, doc_frags: List[OdxDocFragment], *,
+                             internal_type: DataType,
+                             physical_type: DataType) -> "ScaleLinearCompuMethod":
+        cm = CompuMethod.compu_method_from_et(
+            et_element, doc_frags, internal_type=internal_type, physical_type=physical_type)
+        kwargs = dataclass_fields_asdict(cm)
+
+        return ScaleLinearCompuMethod(**kwargs)
+
+    def __post_init__(self) -> None:
+        self._segments: List[LinearSegment] = []
+
+        odxassert(self.category.value == "SCALE-LINEAR",
+                  "ScaleLinearCompuMethod must exibit SCALE-LINEAR category")
+
+        odxassert(self.physical_type in [
+            DataType.A_FLOAT32,
+            DataType.A_FLOAT64,
+            DataType.A_INT32,
+            DataType.A_UINT32,
+        ])
+        odxassert(self.internal_type in [
+            DataType.A_FLOAT32,
+            DataType.A_FLOAT64,
+            DataType.A_INT32,
+            DataType.A_UINT32,
+        ])
+
+        if self.compu_internal_to_phys is None:
+            odxraise("SCALE-LINEAR compu methods require COMPU-INTERNAL-TO-PHYS")
+            return
+
+        compu_scales = self.compu_internal_to_phys.compu_scales
+
+        for scale in compu_scales:
+            self._segments.append(
+                LinearSegment.from_compu_scale(
+                    scale, internal_type=self.internal_type, physical_type=self.physical_type))
+
+    def convert_physical_to_internal(self, physical_value: AtomicOdxType) -> Union[float, int]:
+        applicable_segments = [
+            seg for seg in self._segments if seg.physical_applies(physical_value)
+        ]
+        if not applicable_segments:
+            odxraise(r"No applicable segment for value {physical_value} found")
+            return None
+        elif len(applicable_segments):
+            odxraise(r"Multiple applicable segments for value {physical_value} found")
+
+        seg = applicable_segments[0]
+        return seg.convert_physical_to_internal(physical_value)
+
+    def convert_internal_to_physical(self, internal_value: AtomicOdxType) -> Union[float, int]:
+        applicable_segments = [
+            seg for seg in self._segments if seg.internal_applies(internal_value)
+        ]
+        if not applicable_segments:
+            odxraise(r"No applicable segment for value {internal_value} found")
```

Comment:
What does the specs say about this?

Reply:
we must be able to decode the internal value. probably I should throw a `DecodeError` instead, though. fixed: https://github.com/mercedes-benz/odxtools/pull/306/commits/10bc14b68151debc7084042f17ee86f72a775416